runtime/compiler/optimizer/InterpreterEmulator.hpp

```diff
@@ -0,0 +1,253 @@
+ *        information like known object and constant integer so
+ *        that inliner can make better decision when creating callsites. For
+ *        example, inliner can avoid creating callsites on dead path
+ *        \ref maintainStackForIf or can refine callee method based on known
+ *        receiver info \ref refineResolvedCalleeForInvokestatic. Operands that
+ *        can't be reasoned about are represented by a dummy operand \ref _unknownOperand
+ *        which doesn't carry any extra information. Currently, with state mode only
+ *        supports methodhandle thunk archetypes.
+ */
+
+#ifndef INTERPRETER_EMULATOR_INCL
+#define INTERPRETER_EMULATOR_INCL
+
+#include "il/Block.hpp"
+#include "ilgen/ByteCodeIteratorWithState.hpp"
+#include "ilgen/J9ByteCodeIterator.hpp"
+#include "compile/Compilation.hpp"
+#include "optimizer/Inliner.hpp"
+#include "optimizer/J9Inliner.hpp"
+#include "optimizer/J9EstimateCodeSize.hpp"
+#include "env/TRMemory.hpp"
+
+class IconstOperand;
+class KnownObjOperand;
+class KnownObjPairOperand;
+
+class Operand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      virtual IconstOperand *asIconst(){ return NULL;}
+      virtual KnownObjOperand *asKnownObject(){ return NULL;}
+      virtual KnownObjPairOperand* asKnownObjectPair(){ return NULL;}
+      virtual bool isUnkownOperand(){ return true;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return TR::KnownObjectTable::UNKNOWN;}
+   };
+
+/*
+ * \class KnownOperand
+ *
+ * \brief represent operands that can be reasoned about at compile time
+ */
+class KnownOperand : public Operand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      virtual bool isUnkownOperand(){ return false;}
+   };
+
+class IconstOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      IconstOperand (int x): intValue(x) { }
+      virtual IconstOperand *asIconst() { return this;}
+      int32_t intValue;
+   };
+
+class KnownObjOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      KnownObjOperand(TR::KnownObjectTable::Index koi):knownObjIndex(koi){ }
+      virtual KnownObjOperand *asKnownObject(){ return this;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return knownObjIndex;}
+      TR::KnownObjectTable::Index knownObjIndex;
+   };
+
+/*
+ * \class KnownObjPairOperand
+ *
+ * \brief represents a pair of known objects
+ *
+ * \note This class is used to support mutable callsite because both the methodhandle object
+ *       and the mutable callsite needs to be tracked so that when creating \c TR_J9MutableCallSite
+ *       the mutable callsite object can be set for the callsite even though it's really the
+ *       methodhandle object that's on the operand stack.
+ *
+ * \see getReturnValueForInvokevirtual
+ * \see refineResolvedCalleeForInvokestatic
+ * \see visitInvokestatic
+ */
+class KnownObjPairOperand : public KnownOperand
```

Comment:
Since this inherits doesn't it already inherit the knownObjIndex? If so do we just need a second index? I'm interested why this special case and not some kind of Operand class which can hold a collection of these Operands or some other more generic structure rather than this specializaiton - might we need to know a known object and an index for a varhandle for example?

Reply:
Thanks for the suggestions. This class is used for a very specialized purpose of tracking mutable callsite target but the name was very general. Changed the class name to `MutableCallsiteTargetOperand` as well as the fields' name to be more specific.